AlphaWallet/Foundation/Formatters.swift

```diff
@@ -0,0 +1,103 @@
+//
+//  Formatters.swift
+//  AlphaWallet
+//
+//  Created by Jerome Chan on 17/1/22.
+//
+
+import Foundation
+
+struct Formatters {
+
+    static let currency: NumberFormatter = basicCurrencyFormatter()
+
+    static let usd: NumberFormatter = {
+        let formatter = basicCurrencyFormatter()
+        formatter.positiveFormat = "###,###,###,###,###,###,##0.00" + " " + Constants.Currency.usd
+        formatter.negativeFormat = "-###,###,###,###,###,###,##0.00" + " " + Constants.Currency.usd
+        formatter.currencyCode = ""
+        return formatter
+    }()
+
+    static let percent: NumberFormatter = {
+        let formatter = basicCurrencyFormatter()
+        formatter.positiveFormat = "###,###,###,###,###,###,##0.00"
+        formatter.negativeFormat = "-###,###,###,###,###,###,##0.00"
+        formatter.numberStyle = .percent
+        return formatter
+    }()
+
+    static let shortCrypto: NumberFormatter = {
+        let formatter = basicCurrencyFormatter()
+        formatter.positiveFormat = "###,###,###,###,###,###,##0.0000"
```

Comment:
```swift
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US")
        formatter.positiveFormat = "a ###,###,###,###,###,###,##0.0"
        formatter.negativeFormat = "b-###,###,###,###,###,###,##0.0000"
        print(formatter.maximumFractionDigits)
        print(formatter.minimumFractionDigits)
        print(formatter.string(from:  1.1111)!)
        print(formatter.string(from: -1.1111)!)
```
results in 
```
1
1
a 1.1
b-1.1
```
So it appears that it takes the positiveFormat and uses that for min and max fraction digits.
```swift
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US")
        formatter.positiveFormat = "a ###,###,###,###,###,###,##0.0000"
        formatter.negativeFormat = "b-###,###,###,###,###,###,##0.0"
        print(formatter.maximumFractionDigits)
        print(formatter.minimumFractionDigits)
        print(formatter.string(from:  1000.111)!)
        print(formatter.string(from: -1000.11111)!)
```
results in
```
4
4
a 1,000.1110
b-1,000.1111
```
You can set the min and max fraction digits.
```swift
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US")
        formatter.positiveFormat = "a ###,###,###,###,###,###,##0.0000"
        formatter.negativeFormat = "b-###,###,###,###,###,###,##0.0"
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 4
        print(formatter.maximumFractionDigits)
        print(formatter.minimumFractionDigits)
        print(formatter.string(from:  1000.111)!)
        print(formatter.string(from: -1000.11111)!)
```
results in
```
4
2
a 1,000.111
b-1,000.1111
```
So it appears we should set the min/max fractionDigits.

Reply:
```swift
XCTAssertEqual(Formatters.currency.string(from: 123456789123456789.056), "$123,456,789,123,456,789.05")
```
Is there a rounding issue with Swift literals? Test failed with 
```
("Optional("$123,456,789,123,456,780.00")") is not equal to ("Optional("$123,456,789,123,456,789.05")")
```
@oa-s @hboon

Reply:
How about like this:

```
let formatter = NumberFormatter()
formatter.locale = Locale(identifier: "en_US")
//`negativeFormat` will use this too
formatter.positiveFormat = "###,###,###,###,###,###,##0.0000"
print(formatter.string(from:  1.1111)!)
print(formatter.string(from: -1.1111)!)
print(formatter.string(from:  1.11)!)
print(formatter.string(from: -1.11)!)
print(formatter.string(from:  1.111111)!)
print(formatter.string(from: -1.111111)!)
```

i.e. don't set `negativeFormat` as well as `minimumFractionDigits` and `maximumFractionDigits`. The repetition and inconsistency makes it really easy to get it wrong. The rare comment here helps.

And if this work, replace the positiveFormat with something this (or just maybe even better if we rename `Constants.etherFormatterFractionDigits` and replace it with `"###,###,###,###,###,###,##0.0000"`. Maybe more readible?):

```
"###,###,###,###,###,###,##0.\(String(repeating: "0", count: Constants.etherFormatterFractionDigits))"
```

Reply:
> Is there a rounding issue with Swift literals? Test failed with

@eviltofu not exactly because of literals, but because Swift literals are assumed to be of certain types (in this case `Double` so if we choose a smaller test value, eg. `Formatters.currency.string(from: 123456789.056)`, it will probably work as expected because `Double` can represent it without precision loss.